### PR TITLE
smp: Move for loops to use arch_num_cpus instead of CONFIG_MP_NUM_CPUS

### PR DESCRIPTION
**Pull request by galak**
Change for loops of the form:

for (i = 0; i < CONFIG_MP_NUM_CPUS; i++)
   ...

to

unsigned int num_cpus = arch_num_cpus();
for (i = 0; i < num_cpus; i++)
   ...

We do the call outside of the for loop so that it only happens once, rather than on every iteration.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

### CODE REVIEW

**Review by teburd — CHANGES_REQUESTED** [review]
Looks good, the signed off line in the first commit probably should match the second one though.

**Comment by galak** [comment]
> Looks good, the signed off line in the first commit probably should match the second one though.

Oh, good catch.  ;).

**Review by teburd — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]
LGTM. I wonder if an helper like `FOR_EACH_CPU()` could be more handy here instead of boilerplating.

**Review by MaureenHelm — APPROVED** [review]

File: include/zephyr/arch/arch_inlines.h [hunk 1]
```diff
@@ -24,6 +24,16 @@
...
+#elif defined(CONFIG_NIOS2)
+#include <zephyr/arch/nios2/arch_inlines.h>
+#elif defined(CONFIG_MIPS)
+#include <zephyr/arch/mips/arch_inlines.h>
+#elif defined(CONFIG_ARCH_POSIX)
+#include <zephyr/arch/posix/arch_inlines.h>
+#elif defined(CONFIG_SPARC)
+#include <zephyr/arch/sparc/arch_inlines.h>
+#else
+#error "Unknown Architecture"
```
Comment (thoh-ot) [inline include/zephyr/arch/arch_inlines.h:36]: I know i'm a bit late to the party here, but any thoughts on how this is supposed to work with out-of-tree proprietary architectures (such as the one our company is currently using)?

We don't have SMP support so we don't really care about the CONFIG_MP_MAX_NUM_CPUS, but we do care that we now get an "Unknown Architecture" error.

Could we agree to revert the #error change?

**Pull request merged**